### ISSUE TITLE
[HOLD for payment 2024-07-22] [HOLD for payment 2024-07-17] [$250] Whole chat scrolls when editing a message 

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#41984]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.72-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @Christinadobrzyn 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1715325227361939

## Action Performed:

1. Sign into NewDot on Mac Desktop
2. Open a chat, send a long message
3. Hover over chat > click Menu > Edit Comment
4. Notice the curser is at the bottom of the chat
5. Try to scroll up to the top of the edited chat

## Expected Result:
The edited message should only scroll, the rest of the chat should be anchored to the last message
## Actual Result:
The entire chat thread scrolls instead of just the edited message
## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/f02977e0-53b0-4682-9cdf-abe7797ca6c6

https://github.com/Expensify/App/assets/38435837/2e2daade-3589-40f6-84f6-a6ed10182e42

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010b2b609f33e20856</li>
        <li>Upwork Job ID: 1789856426003279872</li>
        <li>Last Price Increase: 2024-06-03</li>
<li>Automatic offers: </li>
<ul>
<li>ishpaul777 | Reviewer | 102598584</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @Christinadobrzyn / @Christinadobrzyn</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @puneetlath / @Christinadobrzyn</details>

**Comment by Christinadobrzyn** [comment]
I can still reproduce following the steps in the OP - I think the important part is that the message you are editing needs to be very long. You'll notice the whole page scrolls instead of just the edited message - that's the problem. 

https://github.com/Expensify/App/assets/51066321/a9ed3966-8949-4f5a-93e4-8b96cfbf9c6a

Adding external.

**Comment by dsootII** [comment]
Contributor details
Your Expensify account email: dsootsecond@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~013cac504ffac3d4c9

**Comment by QichenZhu** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The whole chat scrolls when scrolling a long message in edit mode.

### What is the root cause of that problem?
This bug is caused by two reasons:
1. The wheel event propagates from the text input to the message list.
2. The CSS property `overscroll-behavior` of the text input is not set. (This only affects situations where the text input is already scrolled to the top or bottom, and the user continues to scroll up or down. However, it may be considered a separate issue. I'm not sure if it should be addressed here.)

### What changes do you think we should make in order to solve the problem?
Listening for the wheel event of the [`textInput`](https://github.com/Expensify/App/blob/90a5028863661d9a31283d0f3832f7b34d5b46cd/src/components/Composer/index.tsx#L82),

```TypeScript
textInput.current.addEventListener('wheel', handleWheel);
```

and stopping the event from propagation in the event handler.

### What alternative solutions did you explore? (Optional)
In addition to the main solution, adding `overscrollBehavior: 'none'` to the [`inputStyleMemo`](https://github.com/Expensify/App/blob/90a5028863661d9a31283d0f3832f7b34d5b46cd/src/components/Composer/index.tsx#L308).

**Comment by aksh1t** [comment]
Contributor details
Your Expensify account email: [akshat.patel16@gmail.com](mailto:akshat.patel16@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~0147e3e82824ea22a4

**Comment by aksh1t** [comment]
Contributor details
Your Expensify account email: akshat.patel16@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0147e3e82824ea22a4

**Comment by shlommo** [comment]
Contributor details
Your Expensify account email: [szhaqypbek@gmail.com](mailto:szhaqypbek@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~015734d9ee71ea225c

**Comment by shlommo** [comment]
Contributor details
Your Expensify account email: szhaqypbek@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~015734d9ee71ea225c

**Comment by Christinadobrzyn** [comment]
@dsootII Welcome! Can you provide a proposal for how to fix this issue, (@QichenZhu created a proposal above, as an example).

**Comment by aksh1t** [comment]
When testing this, I can also reproduce this on MacOS Chrome as well in addition to MacOS Desktop.

**Comment by aksh1t** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

There is unintended dual scrolling when editing a large message in the chat screen. The inner text area (which is being edited) and the outer message list both scroll simultaneously.

This problem happens in both web and desktop. (This is incorrectly mentioned in the issue as only affecting desktop).

### What is the root cause of that problem?

The root cause is the scroll event from the inner text area propagating to the outer list, causing both to scroll at the same time. It should be consumed or blocked by the inner text area.

### What changes do you think we should make in order to solve the problem?

1. Make the `textInput` listen to `wheel` event and call `event.stopPropagation();` to stop the event from affecting the parent scroll view.
2. In `Composer`, set `shouldContainScroll` to `true`. This will prevent overscrolling.

### What alternative solutions did you explore? (Optional)

Explored adding mouse enter and leave listeners to conditionally enable or disable scrolling based on the mouse's position. However, this method adds unneeded complexity.

**Comment by coderdev135** [comment]
Contributor details
Your Expensify account email: coderdev27@protonmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01d03b3b9d60c82dde

**Comment by coderdev135** [comment]
Contributor details
Your Expensify account email: coderdev27@protonmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01d03b3b9d60c82dde

**Comment by coderdev135** [comment]
Contributor details
Your Expensify account email: sourcecodedeveloper@protonmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01d03b3b9d60c82dde

**Comment by coderdev135** [comment]
Contributor details
Your Expensify account email: coderdev27@protonmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01d03b3b9d60c82dde

**Comment by Christinadobrzyn** [comment]
Awesome! Thanks everyone! @ishpaul777 could you check these proposals when you have a moment? Thank you!

**Comment by Christinadobrzyn** [comment]
Awesome! Thanks everyone! @ishpaul777 could you check these proposals when you have a moment? Thank you!

**Comment by QichenZhu** [comment]
I tested both `onWheel` and `addEventListener('wheel', ...)` and found that the former didn't work but the latter worked.

**Comment by aksh1t** [comment]
[Updated Proposal](https://github.com/Expensify/App/issues/41984#issuecomment-2109995201) to add video showing the fixed scrolling.

@dragnoir I can confirm that `addEventListener('wheel', ...)` + `event.stopPropagation();` does indeed fix the scrolling issue. Attached the video which shows the fixed app in my proposal.

**Comment by QichenZhu** [comment]
@ishpaul777 As mentioned in my proposal, there are actually two issues. One issue is wheel event propagation, and another issue is over-scrolling.

You can test my solution to the first issue using [this branch](https://github.com/QichenZhu/App/tree/fix/41984).

Once the first issue is resolved, the second issue will appear. I think @aksh1t proposed a better solution to the second issue.

**Comment by aksh1t** [comment]
@ishpaul777 [Here's the branch](https://github.com/aksh1t/App/tree/akshat/41984-double-scroll-bug) to test the solution.

**Comment by QichenZhu** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/41984#issuecomment-2107122293)

**Comment by QichenZhu** [comment]
@ishpaul777 [Proposal](https://github.com/Expensify/App/issues/41984#issuecomment-2107122293) and [branch](https://github.com/QichenZhu/App/tree/fix/41984) updated to address both issues.

**Comment by aksh1t** [comment]
Hi @ishpaul777 - I have updated both my [proposal](https://github.com/Expensify/App/issues/41984#issuecomment-2109995201) and [branch](https://github.com/aksh1t/App/tree/akshat/41984-double-scroll-bug) to fix the issue. Also updated the video to show it.

@QichenZhu I tested your solution as well, and for me the `isScrollbarVisible` does not seem to allow us to scroll the outer view when the scroll is captured by the inner text field. So, once the inner view "captures" the wheel, the only way to scroll the outer view is to move the mouse pointer out of the view and then scroll.

I managed to fix this by detecting when the inner view is already at the top or bottom position and then in those conditions, not blocking the event propagation. That seems to do the trick.

**Comment by aksh1t** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/41984#issuecomment-2109995201)

**Comment by Christinadobrzyn** [comment]
Melvbot update is: @ishpaul777 is reviewing - thanks!

**Comment by Christinadobrzyn** [comment]
sorry to bother @ishpaul777 but any updates on the review?

**Comment by QichenZhu** [comment]
Thanks for your time. That's weird... Just want to confirm you were testing on my updated solution and branch from https://github.com/Expensify/App/issues/41984#issuecomment-2118627557.

https://github.com/Expensify/App/assets/57348009/8f592424-3356-4906-aa78-79d5d74ebfc6

**Comment by aksh1t** [comment]
@ishpaul777 I tried running my branch again, but was not sure what exactly is the glitch you are talking about. It is not very apparent from the video. Do you mean when the top / bottom is hit, the view drops some frames and stutters? Or does it change direction?

Can you please let me know if you are checking using trackpad on your mac or mouse?

@QichenZhu What he is talking about is that once you are scrolling in your inner text view, you need to move the mouse pointer out of it first and then scroll to scroll the main view. In your video itself at 0:05 seconds, you are moving the mouse out of the view to scroll the bigger view. This should not be the case, we should be able to scroll the bigger view from within the smaller view. For an example of how it should work you can check out how Slack handles it.

**Comment by QichenZhu** [comment]
@aksh1t Thanks for your comment.

> What he is talking about is that once you are scrolling in your inner text view, you need to move the mouse pointer out of it first and then scroll to scroll the main view.

> This should not be the case, we should be able to scroll the bigger view from within the smaller view.

@ishpaul777 Does the above align with what you meant below?

> solution has the same scrollig issues as seen in this video https://github.com/Expensify/App/issues/41984#issuecomment-2118314478

**Comment by aksh1t** [comment]
Thanks for confirming. I'll try to fix the glitch.

**Comment by aksh1t** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/41984#issuecomment-2109995201)

**Comment by aksh1t** [comment]
@ishpaul777 I have closely observed Github's handling of the scroll, and they depend on mouse movements in addition to scroll events to decide whether a scroll event should be "captured". I have tried to emulate it closely and updated my proposal. I am no longer getting the stuttering.

You can test it out on my [updated branch here.](https://github.com/Expensify/App/compare/main...aksh1t:App:akshat/41984-double-scroll-bug)

**Comment by Christinadobrzyn** [comment]
For a MelvBot update, I think we're still reviewing proposals? Does that sound right @ishpaul777? Anything I can do to help?

**Comment by Christinadobrzyn** [comment]
Thanks @ishpaul777! 

Update for Melvbot - we are reviewing the proposals

**Comment by Christinadobrzyn** [comment]
@ishpaul777 can you provide an update when have a moment? Thank you!

**Comment by ikevin127** [comment]
I don't have the capacity to review here but since I was tagged as part of the C+ team, I'd like to mention that since we have touch devices too (android / ios mweb), you might want to apply the addEventListener for both events:

```ts
	// wheel for devices using mouse and touchmove for touch devices
	const events = ['wheel', 'touchmove'];
```

Not sure if this is an issue on mweb, but anyway!

**Comment by ikevin127** [comment]
Cool, well I think based on **C+ Process Doc** `Updated process` > `2.` > `c.`:

> If C+ wants to submit a proposal to fix the issue they do so and tag the internal engineer (by adding “`:ribbon:` `:eyes:` `:ribbon:`") to review the proposal.  All other aspects of working on the issue as the same as other contributors (as of Feb 2024 this is very rare).

So I think this is one of the rare cases where you as C+ are submitting a proposal and the reviewer should be the auto-assigned internal engineer as per the docs.

**Comment by QichenZhu** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/41984#issuecomment-2107122293)

**Comment by QichenZhu** [comment]
@ishpaul777 @puneetlath I found a potential solution by using the existing logic in the `react-native-web` instead of reinventing the wheels. Could you review my [updated proposal](https://github.com/Expensify/App/issues/41984#issuecomment-2107122293)?

**Comment by QichenZhu** [comment]
Thank you for your clarification. I'm just confused about whether the chat should scroll once the composer scrolls to its limit. But anyway, we don't need to worry about this, as `react-native-web` already has an [implementation](https://github.com/necolas/react-native-web/blob/54c14d64dabd175e8055e1dc92e9196c821f9b7d/packages/react-native-web/src/vendor/react-native/VirtualizedList/index.js#L457), just as I [proposed](https://github.com/Expensify/App/issues/41984#issuecomment-2107122293). The result will be like this:

https://github.com/Expensify/App/assets/57348009/e6b7423f-e9c5-423c-9051-1b209b927dee

**Comment by QichenZhu** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/41984#issuecomment-2107122293)

**Comment by QichenZhu** [comment]
> > I noticed with your solution when the scroll reaches on top or bottom it starts scrolling the chat list, do we have any workaround for this, this is consistently reproducable everytime the text input is scrolled.
> 
> I don't think this is a problem. I understand that mimicking GH or Slack will be more intuitive but I def don't consider this a problem.

I agree. I don't think it's a big problem either, since GitHub handles this quite similarly to my proposal.

https://github.com/Expensify/App/assets/57348009/4ed8410b-41fa-4003-80c0-edfc83d8ed0e

**Comment by Christinadobrzyn** [comment]
Just a heads up - I'm going to be ooo until June 24th. @puneetlath can you watch this as a BZ teammate while I'm away or would you like me to assign someone new?

**Comment by aksh1t** [comment]
Hi @parasharrajat Apologies for the delay in replying back. @ishpaul777 's [last comment](https://github.com/Expensify/App/issues/41984#issuecomment-2138158114) regarding the implementation about using debouncing and the bug where the scroll view goes in the opposite direction is still valid. I haven't had the time to propose a better solution so far.

**Comment by puneetlath** [comment]
Let's do it. Please go ahead.

**Comment by Christinadobrzyn** [comment]
Update for Melvin - monitoring PR- https://github.com/Expensify/App/pull/44354

**Comment by Christinadobrzyn** [comment]
PR in staging! Getting close!

**Comment by Christinadobrzyn** [comment]
preparing for payment

Payouts due:

- [ ] Contributor: $250 @parasharrajat (payment through New Expensify)
- [ ] Contributor+: $250 @ishpaul777 (offer in Upwork - https://www.upwork.com/nx/wm/offer/102598584)

@parasharrajat @ishpaul777 do you think we need a regression test?

**Comment by Christinadobrzyn** [comment]
regression test - https://github.com/Expensify/Expensify/issues/413781

waiting for payment day!

**Comment by Christinadobrzyn** [comment]
Update for Melvin- This is in a holding pattern until the payment date

**Comment by Christinadobrzyn** [comment]
Payment day is upon us! 

No regressions so paying this out based on this [payment summary](https://github.com/Expensify/App/issues/41984#issuecomment-2233842247) 

@parasharrajat is ready for NewDot payment! 

Closing this out.

**Comment by JmillsExpensify** [comment]
$250 approved for @parasharrajat
